### PR TITLE
Generalize catArray for contiguous inputs and dim != 0

### PR DESCRIPTION
**Pull request by jamesr66a**
I noticed that we were sinking a lot of time into `cat` operations in machine translation on CPU, and drilled down to us doing the cat element-by-element, even though all the inputs were contiguous. The reason was we were doing the cat along a dimension that was not 0, and that caused us to not use the fast `memcpy` branch. This PR generalizes that branch.

Quick benchmark script:
```
import torch, time

tensors = [torch.rand(6, 2, 1024) for i in range(5)]

NITER = 1000
s = time.time()
for i in range(NITER):
    torch.cat(tensors, dim=1)
print('time per iter ', (time.time() - s) / NITER)
```

Before:
```
time per iter  8.089399337768554e-05
```

After:
```
time per iter  2.183413505554199e-05
```

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Do our tests explicitly exercise this branch? If not, please add.

**Comment by jamesr66a** [comment]
I believe this test exercises the branch:

https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L4203

**Comment by cpuhrsch** [comment]
Looks like it. Maybe double check and make sure the edge case you mentioned is covered as well.
 
In general we prefer to avoid modifying TH and instead porting over the function to aten. Maybe I could ask you to spend a bit of time on seeing how feasible that is?

**Comment by gchanan** [comment]
I have a slight preference not to port this at the same time as we make changes.  It's harder to review and less obvious where the problem is if there's a bug report.

**Review by gchanan — APPROVED** [review]
